pkg/util/workqueue/metrics.go

```diff
@@ -17,63 +17,108 @@ limitations under the License.
 package workqueue
 
 import (
+	"sync"
 	"time"
-
-	"github.com/prometheus/client_golang/prometheus"
 )
 
+// You need to import the prometheus package to create and register prometheus
+// metrics, otherwise all the operations of the metrics are noop.
+
 type queueMetrics interface {
 	add(item t)
 	get(item t)
 	done(item t)
 }
 
+// GaugeMetric that represents a single numerical value that can
+// arbitrarily go up and down.
+// Its methods are a subset of prometheus.Gauge.
+type GaugeMetric interface {
+	Inc()
+	Dec()
+}
+
+type noopGaugeMetric struct{}
+
+func (noopGaugeMetric) Inc() {}
+func (noopGaugeMetric) Dec() {}
+
+// CounterMetric represents a single numerical value that only ever
+// goes up.
+// Its methods are a subset of prometheus.Counter.
+type CounterMetric interface {
+	Inc()
+}
+
+type noopCounterMetric struct{}
+
+func (noopCounterMetric) Inc() {}
+
+// SummaryMetric captures individual observations.
+// Its methods are a subset of prometheus.Observe.
+type SummaryMetric interface {
+	Observe(float64)
+}
+
+type noopSummaryMetric struct{}
+
+func (noopSummaryMetric) Observe(float64) {}
+
 type defaultQueueMetrics struct {
-	depth                prometheus.Gauge
-	adds                 prometheus.Counter
-	latency              prometheus.Summary
-	workDuration         prometheus.Summary
+	// current depth of a workqueue
+	depth GaugeMetric
+	// total number of adds handled by a workqueue
+	adds CounterMetric
+	// how long an item stays in a workqueue
+	latency SummaryMetric
+	// how long processing an item from a workqueue takes
+	workDuration         SummaryMetric
 	addTimes             map[t]time.Time
 	processingStartTimes map[t]time.Time
 }
 
-func newQueueMetrics(name string) queueMetrics {
-	var ret *defaultQueueMetrics
-	if len(name) == 0 {
-		return ret
-	}
+// DepthMetricProvider creates DepthMetric.
+type DepthMetricProvider interface {
```

Comment:
These should be condensed to the types of metrics they provide, not what the metrics are used for. i.e.

GaugeMetricProvider
ConterMetricProvider
SummaryMetricProvider